weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -702,16 +704,43 @@ def refs_read_batch(self, req: tsi.RefsReadBatchReq) -> tsi.RefsReadBatchRes:
         if len(req.refs) > 1000:
             raise ValueError("Too many refs")
 
+        # First, parse the refs
         parsed_raw_refs = [refs_internal.parse_internal_uri(r) for r in req.refs]
+
+        # Business logic to ensure that we don't have raw TableRefs (not allowed)
         if any(isinstance(r, refs_internal.InternalTableRef) for r in parsed_raw_refs):
             raise ValueError("Table refs not supported")
-        parsed_refs = typing.cast(
-            typing.List[refs_internal.InternalObjectRef], parsed_raw_refs
-        )
+        parsed_refs = typing.cast(ObjRefListType, parsed_raw_refs)
 
-        root_val_cache: typing.Dict[str, typing.Any] = {}
+        # Next, group the refs by project_id
+        refs_by_project_id: dict[str, ObjRefListType] = defaultdict(list)
+        for ref in parsed_refs:
+            refs_by_project_id[ref.project_id].append(ref)
+
+        # Lookup data for each project, scoped to each project
+        final_result_cache: typing.Dict[str, typing.Any] = {}
 
         def make_ref_cache_key(ref: refs_internal.InternalObjectRef) -> str:
+            return ref.uri()
+
+        for project in refs_by_project_id:
+            project_refs = refs_by_project_id[project]
+            project_results = self._refs_read_batch_within_project(
+                project, refs_by_project_id[project]
+            )
+            for ref, result in zip(project_refs, project_results):
+                final_result_cache[make_ref_cache_key(ref)] = result
+
+        # Return the final data payload
+        vals = [final_result_cache[make_ref_cache_key(ref)] for ref in parsed_refs]
+        return tsi.RefsReadBatchRes(vals=vals)
+
+    def _refs_read_batch_within_project(
+        self, project_id_scope: str, parsed_refs: ObjRefListType
```

Comment:
not 100% sure what the value is in changing `project_id` to `project_id_scope`

Reply:
mainly for code readability - basically `project_id` is part of a `ref`, so it was a little confusing. `project_id_scope` just is a more clear noun